core/services/ocr2/plugins/ocr2keeper/evmregistry/v21/logprovider/buffer_v2.go

```diff
@@ -0,0 +1,324 @@
+	for len(result) < maxResults && remaining > 0 {
+		nextResults, nextRemaining := b.tryDequeue(start, end, upkeepLimit, maxResults-len(result), upkeepSelector)
+		result = append(result, nextResults...)
+		remaining = nextRemaining
+	}
+
+	return result, remaining
+}
+
+func (b *logBuffer) tryDequeue(start, end int64, minUpkeepLogs, capacity int, upkeepSelector func(id *big.Int) bool) ([]BufferedLog, int) {
+	var result []BufferedLog
+	var remainingLogs int
+	for _, buf := range b.upkeepBuffers {
+		if !upkeepSelector(buf.id) {
+			continue
+		}
+		if capacity == 0 {
+			// if there is no more capacity for results, just count the remaining logs
+			remainingLogs += buf.size()
+			continue
+		}
+		if minUpkeepLogs > capacity {
+			// if there are more logs to fetch than the capacity, fetch the minimum
+			minUpkeepLogs = capacity
+		}
+		logs, remaining := buf.dequeue(start, end, minUpkeepLogs)
+		for _, l := range logs {
+			result = append(result, BufferedLog{ID: buf.id, Log: l})
+			capacity--
+		}
+		remainingLogs += remaining
+	}
+	return result, remainingLogs
+}
+
+func (b *logBuffer) getUpkeepBuffer(uid *big.Int) (*upkeepLogBuffer, bool) {
+	b.lock.RLock()
+	defer b.lock.RUnlock()
+
+	ub, ok := b.upkeepBuffers[uid.String()]
+	return ub, ok
+}
+
+func (b *logBuffer) setUpkeepBuffer(uid *big.Int, buf *upkeepLogBuffer) {
+	b.lock.Lock()
+	defer b.lock.Unlock()
+
+	b.upkeepBuffers[uid.String()] = buf
+}
+
+type upkeepLogBuffer struct {
+	lggr logger.Logger
+
+	id      *big.Int
+	maxLogs *atomic.Int32
+
+	q       []logpoller.Log
+	visited map[string]int64
+	lock    sync.RWMutex
+}
+
+func newUpkeepLogBuffer(lggr logger.Logger, id *big.Int, maxLogs int) *upkeepLogBuffer {
+	limit := &atomic.Int32{}
+	limit.Add(int32(maxLogs))
+	return &upkeepLogBuffer{
+		lggr:    lggr.With("id", id.String()),
+		id:      id,
+		maxLogs: limit,
+		q:       make([]logpoller.Log, 0, maxLogs),
+		visited: make(map[string]int64),
+	}
+}
+
+func (ub *upkeepLogBuffer) setConfig(maxLogs int) {
+	ub.maxLogs.Store(int32(maxLogs))
+}
+
+func (ub *upkeepLogBuffer) size() int {
+	ub.lock.RLock()
+	defer ub.lock.RUnlock()
+
+	return len(ub.q)
+}
+
+func (ub *upkeepLogBuffer) dequeue(start, end int64, limit int) ([]logpoller.Log, int) {
+	ub.lock.Lock()
+	defer ub.lock.Unlock()
+
+	if len(ub.q) == 0 {
+		return nil, 0
+	}
+
+	var results []logpoller.Log
+	var remaining int
+	updatedLogs := make([]logpoller.Log, 0, ub.maxLogs.Load())
+	for _, l := range ub.q {
+		if l.BlockNumber >= start && l.BlockNumber <= end {
+			if len(results) < limit {
+				results = append(results, l)
+				continue
+			}
+			remaining++
+		}
+		updatedLogs = append(updatedLogs, l)
+	}
+
+	if len(results) > 0 {
+		ub.q = updatedLogs
+	}
+
+	prommetrics.AutomationLogsInLogBuffer.Sub(float64(len(results)))
+
+	return results, remaining
+}
+
+func (ub *upkeepLogBuffer) enqueue(offsetBlock int64, logsToAdd ...logpoller.Log) (int, int) {
```

Comment:
Is `baseBlock` or `blockThreshold` a better name than `offsetBlock` for what we're using it to check against?